### PR TITLE
Missing ITs for browse by authority values fix

### LINKED ISSUE #2798: DS-4270 Missing ITs for the browse by authority values
This is a duplicate of [DS-4270](https://jira.lyrasis.org/browse/DS-4270)

There are not test yet to validate the implementation of the browse by metadata when an authority is present. The only test are based on textual values (plain subjects / authors)

This is related to https://github.com/DSpace/Rest7Contract/pull/128

### LINKED ISSUE #3185: Facet entries in the discover/facets/{facet} endpoint does not return the authority key
**Describe the bug**
When calling the {dspace.url}/api/discover/facets/{facet} endpoint, the results won't include the authority key, even though it should be there.

**To Reproduce**
1. Set up a discovery core that has authority values for facets in the config
2. Try to retrieve the values through the call above
3. The resulting facet entries won't include the authority key

**Expected behavior**
The authority keys should be present in the response

**Related work**
This was discovered due to the failing tests in this PR: https://github.com/DSpace/DSpace/pull/3176

### PR DESCRIPTION
**Pull request by Raf-atmire**
## References
* Fixes #2798
* Fixes #3185

## Description
This PR ensures that there are ITs for authority values on the browse endpoint and the facet endpoint

## Instructions for Reviewers

List of changes in this PR:
* Added a way to clear the cache in MetadataAuthorityService as we need to be able to reload the config in those maps during the tests or we won't be able to use specific config in the test cases
* Added a cleanupCache method to the AbstractIntegrationTestWithDatabase class to ensure that the caches of those services are actually cleared properly. The config properties are set to default again, but we had to hard reset the cache of the service as well as to not bring test-specific config to other test cases
* Added the tests, however the Facet test doesn't work as AuthorityValues aren't supported there

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by Raf-atmire** [comment]
Hi @tdonohue ! This is currently by design, this PR simply adds the ITs that actually test for those authority values. Currently the discovery facets endpoints (the test that is failing) doesn't give its authority values in the response and therefore the test fails, indicating that this functionality needs to be fixed.

**Comment by tdonohue** [comment]
@Raf-atmire : Are you planning on fixing the functionality as part of this PR?  If not, can you please create a ticket to describe the bug, linking back to this PR so that we can get someone else assigned to fix the bug?

Obviously, this PR cannot be merged until tests pass.  So, that would mean either we fix the issue, or we temporarily disable (`@Ignore`) that failing test, linking it up to a ticket which need to be resolved before we can re-enable the test.

**Comment by tdonohue** [comment]
@Raf-atmire : In today's meeting, #3177 was assigned to @abollini's team.  So, since you have a test that proves that bug exists, I'd request that you just temporarily disable that test in this PR & add a TODO like `// TODO : should be re-enabled when this issues is fixed: https://github.com/DSpace/DSpace/issues/3177`   That will allow us to review & merge this PR, and then @abollini's team can re-enable the broken test in the PR that fixes the bug.

**Comment by benbosman** [comment]
@tdonohue I just discussed this with Raf, and the failing test is apparently not caused by https://github.com/DSpace/DSpace/issues/3177, but rather by https://github.com/DSpace/DSpace/issues/3185

Do you see any issues in changing the behavior as mentioned in https://github.com/DSpace/DSpace/issues/3185?

This also implies https://github.com/DSpace/DSpace/issues/3177 needs separate tests

File: dspace-api/src/test/java/org/dspace/AbstractIntegrationTestWithDatabase.java [hunk 1]
```diff
@@ -206,6 +209,16 @@ public void destroy() throws Exception {
...
     }
 
+    /**
+     * Clean up service caches after tests
+     * @param serviceManager
+     */
+    private void cleanupCache(ServiceManager serviceManager) {
+        MetadataAuthorityServiceImpl metadataAuthorityService = serviceManager
+            .getServiceByName(null, MetadataAuthorityServiceImpl.class);
+        metadataAuthorityService.clearCache();
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/AbstractIntegrationTestWithDatabase.java:219]: It seems odd to me that we'd need to have this cache cleared _after every test_ (in the destroy() method of this class).  I completely understand needing this cache to be cleared for Authority specific tests, like in the new code you've added to `DiscoveryRestControllerIT`.  Could you describe why this needs to be cleared *both* globally (for every IT test run) and within specific tests (like `DiscoveryRestControllerIT.discoverFacetsAuthorWithAuthorityWithSizeParameter()`? 

It seems like we should only need one or the other, so it's unclear why we need both....and if we don't need both, I'd prefer to remove this global `cleanupCache()` method in favor of clearing the cache only when necessary for the test.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java [hunk 2]
```diff
@@ -215,6 +223,108 @@ public void findBrowseBySubjectEntries() throws Exception {
...
+                   .andExpect(content().contentType(contentType))
+                   .andExpect(jsonPath("$.page.size", is(20)))
+                   //Check that there are indeed 3 different subjects
+                   .andExpect(jsonPath("$.page.totalElements", is(3)))
+                   //Check the embedded resources and that they're sorted alphabetically
+                   //Check that the subject matches as expected
+                   //Verify that they're sorted alphabetically
+                   .andExpect(
+                       jsonPath("$._embedded.entries",
+                            containsInAnyOrder(
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java:294]: In the comment above you mention that the embedded resources should be "sorted alphabetically"...but here you are using `containsInAnyOrder` which doesn't check sort order obviously.  If they are expected in a specific order, we should verify the exact order as well.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java [hunk 3]
```diff
@@ -215,6 +223,108 @@ public void findBrowseBySubjectEntries() throws Exception {
...
+                   .andExpect(content().contentType(contentType))
+                   .andExpect(jsonPath("$.page.size", is(20)))
+                   //Check that there are indeed 3 different subjects
+                   .andExpect(jsonPath("$.page.totalElements", is(3)))
+                   //Check the embedded resources and that they're sorted alphabetically
+                   //Check that the subject matches as expected
+                   //Verify that they're sorted alphabetically
+                   .andExpect(
+                       jsonPath("$._embedded.entries",
+                           containsInAnyOrder(
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java:319]: Same here, we aren't checking the sort order even though we claim to be.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Raf-atmire : Overall, this PR looks good.  But, I had a few inline questions on the code itself that I'd like answered or fixed before we'd merge this.  Thanks!

**Comment by Raf-atmire** [comment]
Hello @tdonohue Thank you for the feedback. I've addressed the issues you've mentioned with the alphabetical sorting and made sure that that's properly checked.
I've also removed the cleanup from the general AbstractIntegrationTestWithDatabase class and added it to the method itself. Please do note that this now includes another `reloadConfig()` call, but this is needed since the destroy of the Items through the ItemBuilders will actually reinstate the cache inside the MetadataAuthorityService. Therefore we need to explicitly reload the config and clear the cache before (or after, but that'd main calling this in the AbstractIntegrationTestWithDatabase) the builders have destroyed their Items, otherwise the config will stay cached inside the service and cause other tests to break.
Could you please perform another review when you get a chance? Thanks!

**Review by tdonohue — APPROVED** [review]
👍 @Raf-atmire : Thanks for the updates here.  I've re-reviewed and this now looks good to me.  @abollini , let me know if you would like to review (it is almost all ITs with a few minor changes to code), or if we should merge with 1 approval.

**Review by abollini — APPROVED** [review]
thanks @Raf-atmire it looks good to me as well with the changes agreed with @tdonohue. I'm going to merge it

**Pull request merged**